OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,220 @@
+ */
+
+package org.sufficientlysecure.keychain.keyimport;
+
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.sufficientlysecure.keychain.network.OkHttpClientFactory;
+import org.sufficientlysecure.keychain.pgp.UncachedKeyRing;
+import org.sufficientlysecure.keychain.pgp.UncachedPublicKey;
+import org.sufficientlysecure.keychain.pgp.exception.PgpGeneralException;
+import org.sufficientlysecure.keychain.ui.util.KeyFormattingUtils;
+import org.sufficientlysecure.keychain.util.ParcelableProxy;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.Proxy;
+import java.net.URL;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import timber.log.Timber;
+
+
+public class WebKeyDirectoryClient implements KeyserverClient {
+
+    public static WebKeyDirectoryClient getInstance() {
+        return new WebKeyDirectoryClient();
+    }
+
+    private WebKeyDirectoryClient() { }
+
+    private static Pattern EMAIL_PATTERN = Pattern.compile("^\\s*(.+)@(.+)\\s*$");
+
+    @Override
+    public List<ImportKeysListEntry> search(String name, ParcelableProxy proxy)
+            throws QueryFailedException, QueryNeedsRepairException {
+        URL webKeyDirectoryURL = toWebKeyDirectoryURL(name);
+
+        if (webKeyDirectoryURL == null){
+            Timber.d("Name not supported by Web Key Directory Client: " + name);
+            return Collections.emptyList();
+        }
+
+        Timber.d("Web Key Directory import: " + name + " using Proxy: " + proxy.getProxy());
+        List<ImportKeysListEntry> entry = new ArrayList<>(1);
+
+        byte[] data = query(webKeyDirectoryURL, proxy);
+        // if we're here that means key retrieval succeeded,
+        // would have thrown an exception otherwise
+        try {
+            try {
+                entry.add(getEntry(data));
+            } catch (UnsupportedOperationException e) {
+                Timber.e("Parsing retrieved Web Key Directory key failed!");
+            }
+        } catch (PgpGeneralException | IOException e) {
+            Timber.e(e, "Failed parsing key from Web Key Directory during search");
+            throw new QueryFailedException("No valid key found on Web Key Directory");
+        }
+        return entry;
+    }
+
+    @Override
+    public String get(String name, ParcelableProxy proxy) throws QueryFailedException {
+        throw new UnsupportedOperationException("Returning armored key from Web Key Directory not supported");
+    }
+
+    private byte[] query(URL url, ParcelableProxy proxy) throws QueryFailedException {
+        try {
+            Timber.d("fetching from Web Key Directory with: " + url + " proxy: " + proxy.getProxy());
```

Comment:
Use interpolation instead of string concatenation:

```
Timber.d("fetching from Web Key Directory with: %s proxy: %s, url, proxy.getProxy());
```

(applies to some other places as well)